src/mon/Monitor.cc

```diff
@@ -2004,9 +2004,17 @@ void Monitor::handle_probe_reply(MonOpRequestRef op)
 			       !has_ever_joined)) {
       dout(10) << " got newer/committed monmap epoch " << newmap->get_epoch()
 	       << ", mine was " << monmap->get_epoch() << dendl;
+      int epoch_diff = newmap->get_epoch() - monmap->get_epoch();
       delete newmap;
       monmap->decode(m->monmap_bl);
-      notify_new_monmap(false);
+      dout(20) << "has_ever_joined: " << has_ever_joined << dendl;
+      if (epoch_diff == 1 && has_ever_joined) {
+        notify_new_monmap(false);
+      } else {
+        notify_new_monmap(false, false);
+      }
+
+      elector.notify_clear_peer_state();
```

Comment:
If we're unconditionally clearing the state, there's not much point to the dance trying to identify the discontinuity above. Shouldn't we just clear it in the "else" branch above and let it be otherwise, though?

Reply:
Addressed in new push